File: Source/Init/BarrierSetup.cs
using System.Collections.Generic;
using WCSharp.Api;
using static WCSharp.Api.Common;

public static class BarrierSetup
{
    private static int BARRIERID = FourCC("YTpb"); // changed to call native only once
    private static bool BarrierActive { get; set; }
    private static unit DummyUnitOne;
    private static unit DummyUnitTwo;
    private static unit DummyUnitThree;
    public static List<destructable> destructables = new();

    public static void Initialize()
    {
        ActivateBarrier();
    }

    private static void CreateDummyUnits()
    {
        var neutralPassive = player.NeutralPassive;
        DummyUnitOne = unit.Create(neutralPassive, Constants.UNIT_DUMMY_1, Regions.Dummy1Region.Center.X, Regions.Dummy1Region.Center.Y, 360);
        DummyUnitTwo = unit.Create(neutralPassive, Constants.UNIT_DUMMY_2, Regions.Dummy2Region.Center.X, Regions.Dummy2Region.Center.Y, 360);
        DummyUnitThree = unit.Create(neutralPassive, Constants.UNIT_DUMMY_3, Regions.Dummy3Region.Center.X, Regions.Dummy3Region.Center.Y, 360);
    }

    private static void CreateBarrier()
    {
        var barrierRegion = Regions.BarrierRegion;
        var centerX = barrierRegion.Center.X;
        var minY = barrierRegion.Rect.MinY;
        var maxY = barrierRegion.Rect.MaxY;

        float distanceY = maxY - minY;
        float intervalY = distanceY / 13;
        for (int i = 1; i < 13; i++)
        {
            float currentY = minY + (i * intervalY);
            var des = destructable.Create(BARRIERID, centerX, currentY);
            destructables.Add(des);
        }
    }

    public static void ActivateBarrier()
    {
        if (BarrierActive) return;
        CreateDummyUnits();
        CreateBarrier();
        BarrierActive = true;
    }

    public static void DeactivateBarrier()
    {
        if (!BarrierActive) return;
        DummyUnitOne.Dispose();
        DummyUnitTwo.Dispose();
        DummyUnitThree.Dispose();
        foreach (var des in destructables)
        {
            des.Dispose();
        }
        destructables.Clear();
        BarrierActive = false;
    }
}

File: src/Init/BarrierSetup.ts
export class BarrierSetup {
    private static BARRIERID: number = FourCC('YTpb') // changed to call native only once
    private static BarrierActive: boolean
    private static DummyUnitOne: Unit
    private static DummyUnitTwo: Unit
    private static DummyUnitThree: Unit
    public static destructables: destructable[] = []

    public static Initialize() {
        ActivateBarrier()
    }

    private static CreateDummyUnits() {
        let neutralPassive = player.NeutralPassive
        DummyUnitOne = Unit.create(
            neutralPassive,
            Constants.UNIT_DUMMY_1,
            Regions.Dummy1Region.Center.x,
            Regions.Dummy1Region.Center.y,
            360
        )
        DummyUnitTwo = Unit.create(
            neutralPassive,
            Constants.UNIT_DUMMY_2,
            Regions.Dummy2Region.Center.x,
            Regions.Dummy2Region.Center.y,
            360
        )
        DummyUnitThree = Unit.create(
            neutralPassive,
            Constants.UNIT_DUMMY_3,
            Regions.Dummy3Region.Center.x,
            Regions.Dummy3Region.Center.y,
            360
        )
    }

    private static CreateBarrier() {
        let barrierRegion = Regions.BarrierRegion
        let centerX = barrierRegion.Center.x
        let minY = barrierRegion.Rect.minY
        let maxY = barrierRegion.Rect.maxY

        let distanceY: number = maxY - minY
        let intervalY: number = distanceY / 13
        for (let i: number = 1; i < 13; i++) {
            let currentY: number = minY + i * intervalY
            let des = destructable.Create(BARRIERID, centerX, currentY)
            destructables.push(des)
        }
    }

    public static ActivateBarrier() {
        if (BarrierActive) return
        CreateDummyUnits()
        CreateBarrier()
        BarrierActive = true
    }

    public static DeactivateBarrier() {
        if (!BarrierActive) return
        DummyUnitOne.dispose()
        DummyUnitTwo.dispose()
        DummyUnitThree.dispose()
        for (let des in destructables) {
            des.dispose()
        }
        destructables.clear()
        BarrierActive = false
    }
}
